### ISSUE TITLE
Subtleties of the path data grammar

### ISSUE THREAD

**Issue by dalboris** [w3c/svgwg#752]
As a lot of us know, the path data grammar is a little convoluted for the sake of path data conciseness.

For example, the spec makes it clear that something like "M0.6.5L100-200" is an intentionally valid path data, equivalent to "M 0.6 0.5 L 100 -200".

In this GitHub issue, I'd like to dig deeper in the rabbit hole, by giving more examples of contrived (or not so contrived) examples, and asking the following three questions for each:

1. Is it currently formally allowed by the grammar?
2. Is it intentionally allowed (or disallowed)?
3. If answers to 2. is "no", should we change the grammar to reflect intent?

I'll just start with a few examples about whitespaces, and we can add more later. I name the examples so we can reference them. This GitHub issue could be used as clarification of intent.

References:

SVG 1.1 grammar: https://www.w3.org/TR/SVG11/paths.html#PathDataBNF
SVG 2-draft grammar: https://svgwg.org/svg2-draft/paths.html#PathDataBNF

# Examples about whitespaces (W)

## Example W1 - trailing whitespaces after path

```
"M1 2L3 4 "
         ^
```

**1. Is it currently formally allowed by the grammar?**

I believe it is allowed by SVG 1.1 grammar, but forbidden by the SVG 2 draft. Relevant parts of the grammar:

```
SVG 1.1      svg-path ::= wsp* moveto-drawto-command-groups? wsp*
SVG 2-draft  svg_path ::= wsp* moveto? (moveto drawto_command*)?
```

(of course, one has to dig deeper into SVG 2-draft's definition to conclude)

**2. Is it intentionally disallowed in SVG 2?**

I hope not. Whitespaces can be used if we prefer to place the closing `"` on its own line, for example, for manually edited SVGs. Besides, leading whitespaces are allowed, so this would be inconsistent.

**3. Should we change the grammar to reflect intent?**

I personally think we should.

## Example W2 - trailing whitespaces after command

```
"M1 2 L3 4"
     ^
```

**1. Is it currently formally allowed by the grammar?**

I believe it is allowed by SVG 1.1 grammar, but forbidden by the SVG 2 draft. Relevant parts of the grammar:

```
SVG 1.1      moveto-drawto-command-group ::= moveto wsp* drawto-commands?
SVG 2-draft  svg_path ::= wsp* moveto? (moveto drawto_command*)?
```

**2. Is it intentionally disallowed in SVG 2?**

I believe this must be an accidental oversight, since even the [first example](https://svgwg.org/svg2-draft/paths.html#PathDataGeneralInformation) of path data contains those whitespaces: `"M 100 100 L 300 100 L 200 300 z"`,

**3. Should we change the grammar to reflect intent?**

Obviously, we should. Together with W1, I believe we should have:

```
path ::= wsp* (moveto wsp* commands)? wsp*
commands ::= command | (command wsp* commands)
```

That is, reverting back to something more similar to SVG 1.1 (but still slightly simpler), and removing from SVG 2 the initial `moveto?` which is completely redundant with the one within the `(moveto drawto_command*)?` group, unless I'm missing something.

## Example W3 - optional whitespaces between arguments

`"M12L34"`

**1. Is it currently formally allowed by the grammar?**

Yes and no.

Formally, it is allowed by the [E]BNF in either SVG 1.1 or SVG 2, since whitespaces are almost always optional between argument (the only exception is before the first `flag` of an elliptical arc, where it is mandatory). So this example can be parsed as `"M 1 2 L 3 4"`.

However, the prose below says:

> The processing of the EBNF **must** consume as much of a given EBNF production as possible, stopping at the point when a character is encountered which no longer satisfies the production.
>
> (emphasis mine) 

which means that when parsing the number after the L, we are not allowed to stop at `1`, we must continue and also consume the `2`. So the first number becomes `12`, and the lack of a second number makes the path invalid.

**2. What's the intent?**

I believe the intent is that `"M12L34"` should be invalid, of which I would whole-fully agree. 

Although interestingly, this example is unambiguous: there is only one valid way the syntax would make it valid, which is "M 1 2 L 3 4". The same is not true for `"M123L45"`, which could be *in theory* parsed as either `"M 1 23 L 4 5"` or `"M 12 3  L 4 5"`.

This is because the EBNF is in fact ambiguous as is, and only disambiguated via the prose, mandating parsers to be "greedy".

**3. Should we change the grammar to reflect intent?**

Changing the grammar to formally allow `"M0.6.5"` but disallow `"M12"` would be extremely difficult and make the grammar quite unreadable, I believe. Which I guess is why we have the prose "clarification".

However, this whole grammar ambiguity is because we allowed things like `"M0.6.5"`in the first place. I would personally advocate to force the presence of comma_wsp at least between command arguments (to make the grammar unambiguous):

```
"M0.6.5L100-200"   =>   "M0.6 .5L100 -200"
```

But also force the presence of whitespaces between commands and between a command character and its arguments (to make it I believe more consistent with CSS tokenizer -- to be confirmed by someone who knows better):

```
"M0.6 .5L100 -200"   =>   "M 0.6 .5 L 100 -200"
```

This is obviously a backward incompatible change, like the one of disallowing the trailing decimal `"nn."`. However, we could simply note in the documentation that it has been disallowed in SVG 2 since the importance of data compactness isn't as important today as it was in the early days of SVG, and that consistency with CSS parsing rules, as well as making the formal grammar unambiguous, was deemed more important.

It's a rather trivial change for tooling generating SVG 2 documents to force the generation of these whitespaces. And obviously, they already know how to parse them.

Any thoughts?

**Comment by BigBadaboom** [comment]
There has not been any intentional change to the grammar to disallow anything that was allowed before.  Any incompatibilities will just be errors that crept in when the grammar was rewritten.

So the answer to all Q2s will be "to stay the same as SVG 1.1", and the answer to all Q3s is "If the grammar says something different from SVG 1.1, then yes it should be fixed.

**Comment by dalboris** [comment]
(Note: at least one thing *was* intentionally disallowed: trailing decimals in numbers, e.g., "42." is allowed in SVG 1.1 but disallowed in SVG 2)

Alright, then I'm 99% sure the first rule should be fixed from:

```
svg_path ::= wsp* moveto? (moveto drawto_command*)?
```

to:

```
svg_path ::= wsp* (moveto (wsp* drawto_command)*)? wsp*
```

There are three atomic changes here:
1. We remove the initial `moveto?`, because it's completely useless, that is, removing it from the grammar still describes the same language.
2. We change `drawto_command*` to `(wsp* drawto_command)*`, so that whitespaces are allowed between commands
3. We add a final `wsp*` so that trailing whitespaces are allowed.

(By "whitespaces", I obviously mean any of the four whitespace characters)

Also, `drawto_command` should really be renamed `command`, since it also includes `moveto` as potential command. The current name is just plain confusing, especially since in SVG 1.1 `drawto-command` was indeed referring to any command *except* `moveto`. In conclusion, we should most definitely change from:

```
svg_path::= wsp* moveto? (moveto drawto_command*)?
drawto_command::=
    moveto
    | closepath
    | lineto
    | horizontal_lineto
    | vertical_lineto
    | curveto
    | smooth_curveto
    | quadratic_bezier_curveto
    | smooth_quadratic_bezier_curveto
    | elliptical_arc
```

to:

```
svg_path::= wsp* (moveto (wsp* command)*)? wsp*
command::=
    moveto
    | closepath
    | lineto
    | horizontal_lineto
    | vertical_lineto
    | curveto
    | smooth_curveto
    | quadratic_bezier_curveto
    | smooth_quadratic_bezier_curveto
    | elliptical_arc
```

**Comment by Zambonifofex** [comment]
Hello. I am not related to spec‐writing in any way, but I feel like I think the general rationale is to carefully investigate the percentage of uses of a feature, before deciding whether it’s appropriate to drop it.

I think the trailing dot is rarely (if ever at all) useful to anyone in anyway, and as a consequence is rarely (if ever at all) used in the Web.

On the other hand, various minification tools employ the ability to remove whitespace between arguments, which makes it frequent in the Web — changing it would be a major breaking change in terms of “percentage of SVG images broken”. You could argue that this kind of minification is needless, and I wouldn’t fully disagree, but at the same time, it has already been too widely used to be feasibly removed.

**Comment by dalboris** [comment]
I completely agree with the fact that optional whitespaces are widely used. In fact, I'd say that SVG images containing them is probably the majority.

But I'd argue that minifiers use them "just because they can", not out of real necessity.

Also, like trailing decimal, we could add a note to the spec saying that user agents may still accept those. And more to the point, if they choose to be strict and reject them, they should only do so for images explicitly marked as SVG 2.x, so it wouldn't break any existing content. Tools generating SVG 2.x content would trivially update their code to make spaces mandatory (it's less code to make them mandatory than optional).

In conclusion, making whitespaces mandatory in the spec, while having UAs be robust in case of missing whitespaces, could be a smooth long term transition to having getting rid of them without breaking anything, and eventually be able to unify CSS and path-data syntax.

What I see as an important question, however, is: are there use cases where the gain of compactness actually matters to the point where some authors would genuinely miss the feature? I think optional whitespaces can somesave up to 25% of file size in an SVG, which isn't nothing.This is harder to answer, and I think not knowing is what would make me hesitant to make such change.
New content generetors(e.g., minifiers) wanting to createwould be invalid. Like for
However, I'm not sure isn't a useful "feature" as in: "people would miss it"

**Comment by dalboris** [comment]
(sorry about the incomplete message and close/open... I missclicked on the close button, on mobile, please see edit)

**Comment by tatarize** [comment]
From the programming perspective it's generally irrelevant. You pattern search for a letter that isn't an 'e' in the path, then you split the data at that point. All numbers and non-letter (except 'e') symbols between that letter and the next letter are part of that command. You then parse that command individually. You get the command operands by running a greedy regex match across the data and finding everything that matches the generally accepted float regex. `[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?`

Or at least I thought that until I ran into `M200,120 h-25 a25,25 0 1125,25 z` and raised issue #755 . The truth is regex is so ubiquitous that that is going to be how most everybody parses that grammar. Though throwing it into LEX could certainly be a thing.

For the most part since I'm going to fed it all to a greedy regex anyway and cast it to float. The syntax doesn't matter that much, just that it's compact and easy to feed to well established parsers.  My objection with the arc flags is that it breaks that pattern. There's valid svg path patterns that I'm not going to rewrite everything just to support because you could legally remove some spaces that made suddenly makes the context matter. Because I'm in an A value at the 4th position 1125 parsed as 1 1 25 rather than 1125 which is what it would be everywhere else. That sort of context sensitive parsing is horrible.

1) M 12 34 could be taken as M 1,2 L 3, 4 or M 12,34 without greedy parsing. Solving M 123 L... would be impossible and we'd have to detect the exactly two characters things out of the gate. If we have M12 as our path already and append a point to that path, we shouldn't have to care where we came from. But, without greedy parsing we'd need to know we didn't perform one of these unambiguous correction things. Also, it's a bit out there, but if somebody wanted to adapt pathing to 3d or 1d not discretizing the operands will cause ambiguity. M 12 34 in a 1d path makes sense. Go to position (12) then to position (34) it's a line (as are all 1d things). D1 M12 34 (with D being a hypothetical dimensionalization command) would be unambiguous. Also, the general grammar is so nice that fiddling with it like that seems to be a non-starter.

2) Greedy parsing is better parsing. Because then I don't need to consider the context. I simply slurp up any logically reasonable character and I'm set. If you have more data, that's fine it won't change the info I've already gobbled up.

3. No. Grammars should not reflect intent. They should be the the expression of what is and isn't permitted by expressions. Is something grammatically valid "Time flies like an arrow. Fruit flies like a banana." is different than is does that make actual sense: "When is the palace starting?" "Who murdered the wind?" -- I think excluding things through mere grammar is a bit strange. You are saying this element is unparsable rather than this element is malformed. `M0,0A -25, 25 0 1 1 50,0` as a path is technically unparsable. The `-` sign there in front of the radius, does not compute with the grammar. We are admonished exactly how to deal with it. Namely absolute value of radius. But, we're supposed to, according to the grammar, not know what that says, a minus there is a syntax error. And I think basically everybody reads it and deals with it correctly.

I would actually prefer if the grammar was made to be *more* general. Namely:

> wsp::= (#x20 | #x9 | #xD | #xA | comment)
> comment::= \([^\)]\)
> non-e-letter::= [abcdfghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ]
> css-number::= [-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?
> path::= (non-e-letter wsp* (wsp* css-number wsp*)*)*

Exclude anything in brackets as being parsed `(` unparsed `)`

Then specify that the letter is a command and the numbers are operands and here's what commands SVG supports and what range or values are supported and how to make out of range values fit these in range values. If another command is expressed it's an error unless some other namespace supports it. 

Now, more to your point. The documentation should outline preferred export syntax. With regards to robustness we should be liberal with what we accept and conservative with what we produce. There I would suggest all point pairs be divided by comma, all other operands be explicitly separated by spaces, we exclusively use absolute positioning, L be preferred over V and H (in absolute they get confusing). T and S be used wherever possible.

d="
(draw a closed semicircle)
M 0,0 (move to origin)
A 25 25 0 1 1 50,0 (semicircle arc)
Z (close to 0,0)
"

**Comment by dalboris** [comment]
@tatarize I agree with your general idea: I find it more elegant to have the *grammar* itself be much simpler as you propose (e.g., not rejecting 1.0 as a flag, not rejecting -1.0 as an unsigned, and not  rejecting an incorrect number of arguments), and instead have the **operators** (M, L, etc...) be defined in such a way that they are robust to incorrect arguments (e.g., "stop processing path" if incorrect number of argument, or "take the absolute value" for arguments which are supposed to be non-negative).

Unfortunately, doing so would indeed be a breaking change, since as you remarked, things like `"...a25,25 0 1125,25 z"` are currently accepted, and couldn't be with a simpler grammar. I'm actually in favor of revamping the path-data syntax in a non-backward compatible way, but in this GitHub issue I wanted mostly to focus on fixing small unintentional errors that creeped in while rewriting the grammar from SVG 1.1 to SVG 2.

Just a few minor notes:

> No. Grammars should not reflect intent. 

I think you misunderstood what I meant by "intent". All I was saying is, for example: "The curent SVG 2 draft does not allow trailing whitespaces, while SVG 1.1 does allow them: is this an **intent**ional change?"

> the truth is regex is so ubiquitous that that is going to be how most everybody parses that grammar.

Actually, this is not how I'm parsing it. I do use a regex for "parsing the next number when I know that a number is expected", but I don't use a regex to split the path into commands. The latter is done with custom-made code, following the SVG spec very precisely: 

https://github.com/dalboris/vpaint/blob/696ba2ecee67f5e7086b32c18cdb15425b0bbe83/src/Gui/SvgParser.cpp#L339

For example, I do stop processing the path and issue an error if the character `+` or `-` is encountered while reading the first or second argument of an elliptical arc.

> The syntax doesn't matter that much, just that it's compact and easy to feed to well established parsers. [...] That sort of context sensitive parsing is horrible.

I agree that it's annoying that parsing arguments is context-sensitive (depends on the current operator and argument position), and it makes writing the parser harder. However, this context-sensitivity does make the syntax more compact, which was I believe the rationale for making this decision back in the days where compactness mattered more than today.

**Comment by tatarize** [comment]
Your methodology at does give solid ideas as to how to properly do that. I could just build overt regexes for all the commands. Though it's decidedly less simple and only would properly parse the arc command, since everything else merely consists of `letter + wsp + (number wsp)*`. It would however also succeed the test where you're required to stop reading the path because they put a `#` in the command. The only gain in compactness is literally in Arc where you can potentially save 2 characters.

**Comment by dalboris** [comment]
> The only gain in compactness is literally in Arc where you can potentially save 2 characters.

Indeed, and saving these two characters is pretty silly knowing that (including these two characters), the 7 arguments of an arc take at least 14 characters, so best case scenario you compress from 14 to 12 characters (14.3% compression), which is a lower compression factor than typical lossless compression you'd achieve by simply zipping the file for transmission.

But oh well... backward compatibility. At least this doesn't impact behavior, it "just" makes it harder to write parsers.

**Comment by dalboris** [comment]
Fixed by #778
